lib/promscrape/scrapework.go

```diff
@@ -527,51 +527,64 @@ func (sw *scrapeWork) processDataOneShot(scrapeTimestamp, realTimestamp int64, b
 }
 
 func (sw *scrapeWork) processDataInStreamMode(scrapeTimestamp, realTimestamp int64, body *bytesutil.ByteBuffer, scrapeDurationSeconds float64) error {
-	samplesScraped := 0
-	samplesPostRelabeling := 0
-	wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
+	var samplesScraped atomic.Int64
```

Comment:
Would it make sense to run a limited list of workers here, wehre each worker would have its own once allocated buffer and `samplesScraped` `samplesDroppedTotal` counters?
Then in Parse callback we can pick not busy worker and do the work. Then - return it back to the pool.
Pros: no atomics, as counters are local to workers. no frequent writeRequestCtxPool calls, as workers are allocated once and disposed when `processDataInStreamMode ` finishes.
Cos: a bit more complicated function implementaition.

Reply:
I think, that current implementation is good enough. It also aligns with other usage of stream.Parse methods. For instance, vmagent/prometheusimport uses temporary PushContext for the same purpose. It's cheap and doesn't have any significant overhead. It also doesn't make code more complex.

 So, it's fine to use atomics and writeRequestCtxPool on each callback call.